### PR TITLE
Refactor more text formatting

### PR DESCRIPTION
**Pull request by ZehMatt**

### CODE REVIEW

File: src/openrct2/ride/Ride.cpp [hunk 1]
```diff
@@ -3193,23 +3194,21 @@ static void ride_entrance_exit_connected(Ride* ride)
...
         if (!exit.isNull() && !ride_entrance_exit_is_reachable(exit))
         {
             // name of ride is parameter of the format string
-            auto ft = Formatter::Common();
             ride->FormatNameTo(ft);
             if (gConfigNotifications.ride_warnings)
             {
-                News::AddItemToQueue(News::ItemType::Ride, STR_EXIT_NOT_CONNECTED, ride->id);
+                News::AddItemToQueue(News::ItemType::Ride, STR_EXIT_NOT_CONNECTED, ride->id, ft);
             }
```
Comment (tupaschoal) [inline src/openrct2/ride/Ride.cpp:3212]: I guess you need to reset the Formatter in each of these cases, as the function does not early return

File: src/openrct2/management/NewsItem.h [hunk 2]
```diff
@@ -291,7 +292,7 @@ namespace News
 
     std::optional<CoordsXYZ> GetSubjectLocation(News::ItemType type, int32_t subject);
 
-    News::Item* AddItemToQueue(News::ItemType type, rct_string_id string_id, uint32_t assoc);
+    News::Item* AddItemToQueue(News::ItemType type, rct_string_id string_id, uint32_t assoc, const Formatter& formatter);
```
Comment (tupaschoal) [inline src/openrct2/management/NewsItem.h:295]:
```suggestion
    News::Item* AddItemToQueue(News::ItemType type, rct_string_id string_id, uint32_t assoc, const Formatter& formatter = Formatter());
```

I guess you need this for now, unless you expect to convert them all into this commit

**Review by tupaschoal — COMMENTED** [review]
Nice start, this will clean up so many global usages, I'm delighted :)

File: src/openrct2/drawing/Text.cpp [hunk 3]
```diff
@@ -128,22 +128,14 @@ void DrawTextBasic(
...
-    const void* args = ft.GetStartBuf();
-    TextPaint textPaint = { colour, FONT_SPRITE_BASE_MEDIUM, underline, alignment };
-    gCurrentFontSpriteBase = FONT_SPRITE_BASE_MEDIUM;
-
-    utf8 buffer[512];
-    format_string(buffer, sizeof(buffer), format, args);
-    gfx_clip_string(buffer, width);
-
-    DrawText(dpi, coords, textPaint, buffer);
+    return DrawTextEllipsised(dpi, coords, width, format, ft.Data(), colour, alignment, underline);
```
[the remark this answers is not in the document]
Comment (duncanspumpkin) [reply src/openrct2/drawing/Text.cpp:138]: No it's passing to the void version of the function. I think it should be the other way round though. We are trying to get rid of the void* versions of the functions.

[the remark this answers is not in the document]
Comment (ZehMatt) [reply src/openrct2/drawing/Text.cpp:138]: He is actually right, it was a mistake from rebasing.

File: src/openrct2/interface/Screenshot.cpp [hunk 4]
```diff
@@ -414,15 +414,15 @@ void screenshot_giant()
         WriteDpiToFile(path->c_str(), &dpi, gPalette);
 
         // Show user that screenshot saved successfully
-        auto ft = Formatter::Common();
+        Formatter ft;
         ft.Add<rct_string_id>(STR_STRING);
         ft.Add<char*>(path_get_filename(path->c_str()));
```
[the remark this answers is not in the document]
Comment (ZehMatt) [reply src/openrct2/interface/Screenshot.cpp:419]: I find this a bit ugly to chain it like this, its barely readable imho.

File: src/openrct2/management/NewsItem.h (hunk 2 above)
Comment (ZehMatt) [reply src/openrct2/management/NewsItem.h:295]: Forgot to amend, I did indeed plan to have all the news use it correctly.

File: src/openrct2/localisation/Localisation.h [hunk 5]
```diff
@@ -82,32 +83,39 @@ extern const rct_string_id DateGameShortMonthNames[MONTH_COUNT];
...
         : StartBuf(buf)
         , CurrentBuf(buf)
     {
     }
 
+    Formatter()
+        : StartBuf(Buffer.data())
+        , CurrentBuf(StartBuf)
+    {
+    }
```
[the remark this answers is not in the document]
Comment (ZehMatt) [reply src/openrct2/localisation/Localisation.h:101]: I will do further cleanups, I'll leave it out for now, the goal is to not pass buffers like this anymore.

File: src/openrct2/management/Marketing.cpp [hunk 6]
```diff
@@ -68,9 +68,9 @@ uint16_t marketing_get_campaign_guest_generation_probability(int32_t campaignTyp
 
 static void marketing_raise_finished_notification(const MarketingCampaign& campaign)
 {
+    Formatter ft;
```
Comment (tupaschoal) [inline src/openrct2/management/Marketing.cpp:71]: You can reduce the scope of this object to be only inside the if

**Review by tupaschoal — COMMENTED** [review]
Looks good now. Are you planning on adding more commits? (If not it would be good to squash some of the commits together)

File: src/openrct2/interface/Screenshot.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (tupaschoal) [reply src/openrct2/interface/Screenshot.cpp:419]: Alright, I'm not attached to it.

**Comment by ZehMatt** [comment]
I think ill do more refactor work in other PRs as this easily piles up on changes.

**Comment by ZehMatt** [comment]
I've squashed a few commits, not planning to do more in this PR, so from my side its ready to go.

**Comment by duncanspumpkin** [comment]
Oh no lots of crashes

**Pull request merged**